src/msg/async/frames_v2.cc

```diff
@@ -469,8 +485,48 @@ std::ostream& operator<<(std::ostream& os, const FrameAssembler& frame_asm) {
   }
   os << "rev1=" << frame_asm.m_is_rev1
      << " rx=" << frame_asm.m_crypto->rx.get()
-     << " tx=" << frame_asm.m_crypto->tx.get();
+     << " tx=" << frame_asm.m_crypto->tx.get()
+     << " comp rx=" << frame_asm.m_compression->rx.get()
+     << " tx=" << frame_asm.m_compression->tx.get()
+     << " compressed=" << frame_asm.is_compressed();
   return os;
 }
 
+void FrameAssembler::asm_compress(bufferlist segment_bls[]) {
+  bufferlist out[m_descs.size()];
+
+  m_compression->tx->reset_handler(m_descs.size(), get_frame_logical_len());
+
+  bool abort = false;
+  for (size_t i = 0; (i < m_descs.size()) && !abort; i++) {
+      abort = !m_compression->tx->compress(segment_bls[i], out[i]);
+  }
+
+  if (!abort) {
+    m_compression->tx->final();
+
+    for (size_t i = 0; i < m_descs.size(); i++) {
+      segment_bls[i].swap(out[i]);
+      m_descs[i].logical_len = segment_bls[i].length();
+    }
+
+    m_flags |= FRAME_EARLY_DATA_COMPRESSED;
+  }
+}
+
+void FrameAssembler::disasm_all_decompress(bufferlist segment_bls[]) const {
+  if (!is_compressed()) {
+    return;
+  }
+
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    bufferlist out;
+    
+    if (!m_compression->rx->decompress(segment_bls[i], out)) {
+      throw FrameError(fmt::format("Segment decompression failed"));
+    }
+    segment_bls[i].swap(out);
+  }
```

Comment:
```suggestion
    bufferlist decompressed;
    if (!m_compression->rx->decompress(segment_bls[i], decompressed)) {
      throw FrameError(fmt::format("Segment decompression failed"));
    }
    segment_bls[i] = std::move(decompressed);
  }
```

for couple reasons:

1. better name
1. more explicit that we don't care about `out` after this.